spring-xd-reactor/src/main/java/org/springframework/xd/reactor/SynchronousDispatcherMessageHandler.java

```diff
@@ -70,53 +73,65 @@ public SynchronousDispatcherMessageHandler(Processor processor) {
         Environment.initializeIfEmpty(); // This by default uses SynchronousDispatcher
         Method method = ReflectionUtils.findMethod(this.reactorProcessor.getClass(), "process", Stream.class);
         this.inputType = ResolvableType.forMethodParameter(method, 0).getNested(2);
-
-        //Stream with a SynchronousDispatcher as this handler is called by Message Listener managed threads
-        this.stream = Streams.broadcast();
-
-        //user defined stream processing
-        Stream<?> outputStream = processor.process(stream);
-
-        //Simple log error handling
-        outputStream.when(Throwable.class, new Consumer<Throwable>() {
-            @Override
-            public void accept(Throwable throwable) {
-                logger.error(throwable);
-            }
-        });
-
-        this.consume = outputStream.consume(new Consumer<Object>() {
-            @Override
-            public void accept(Object outputObject) {
-                if (ClassUtils.isAssignable(Message.class, outputObject.getClass())) {
-                    getOutputChannel().send((Message) outputObject);
-                } else {
-                    //TODO handle copy of header values when possible
-                    getOutputChannel().send(MessageBuilder.withPayload(outputObject).build());
-                }
-                //TODO handle Void
-            }
-        });
-
-        if (logger.isDebugEnabled()) {
-            logger.debug(consume.debug());
-        }
-
     }
 
     @Override
     protected void handleMessageInternal(Message<?> message) throws Exception {
-
+        Broadcaster<Object> broadcasterToUse = getBroadcaster();
         if (ClassUtils.isAssignable(inputType.getRawClass(), message.getClass())) {
-            stream.onNext(message);
+            broadcasterToUse.onNext(message);
         } else if (ClassUtils.isAssignable(inputType.getRawClass(), message.getPayload().getClass())) {
             //TODO handle type conversion of payload to input type if possible
-            stream.onNext(message.getPayload());
+            broadcasterToUse.onNext(message.getPayload());
         }
 
-        if (logger.isDebugEnabled()) {
-            logger.debug(consume.debug());
+        //TODO store Controls per-thread to keep this debugging
+//        if (logger.isDebugEnabled()) {
+//            logger.debug(consume.debug());
+//        }
+    }
+
+
+    private Broadcaster<Object>  getBroadcaster() {
+        long idToUse = Thread.currentThread().getId();
+        Broadcaster<Object> broadcaster = broadcasterMap.get(idToUse);
+        if (broadcaster == null) {
+            Broadcaster<Object> existingBroadcaster = broadcasterMap.putIfAbsent(idToUse, Streams.broadcast());
```

Comment:
Will need to see how we handle the equivalent of a "Subscription", but there is currently a potential leak here, I guess.